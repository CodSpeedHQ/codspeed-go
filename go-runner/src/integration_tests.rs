use itertools::Itertools;
use rstest::rstest;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use tempfile::TempDir;

use crate::results::walltime_results::WalltimeResults;

fn setup_test_project(project_name: &str) -> anyhow::Result<TempDir> {
    let project_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("testdata/projects")
        .join(project_name);
    println!("Project path: {project_path:?}");

    let temp_dir = TempDir::new()?;
    crate::utils::copy_dir_all(&project_path, &temp_dir)?;

    Ok(temp_dir)
}

fn assert_results_snapshots(profile_dir: &Path, project_name: &str) {
    let glob_pattern = profile_dir.join("results");
    if !glob_pattern.exists() {
        eprintln!("No results found for project: {project_name}");
        return;
    }

    let files = std::fs::read_dir(&glob_pattern)
        .expect("Failed to read results directory")
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .map(|path| {
            let file = std::fs::File::open(&path).unwrap();
            serde_json::from_reader::<_, WalltimeResults>(file).unwrap()
        })
        // Ensure we have the correct order for multiple test executables
        .sorted_by_cached_key(|r| {
            r.benchmarks
                .iter()
                .map(|b| b.metadata.name.clone())
                .sorted()
                .join(";")
        })
        .collect::<Vec<_>>();

    for (i, mut content) in files.into_iter().enumerate() {
        content
            .benchmarks
            .sort_by_cached_key(|b| b.metadata.name.clone());

        let _guard = {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_suffix(format!("{project_name}_{i}"));
            settings.bind_to_scope()
        };

        insta::assert_json_snapshot!(content, {
            ".creator.pid" => "[pid]",
            ".benchmarks[].stats" => "[stats]",
        });
    }
}

#[rstest]
// // #[case::caddy("caddy")]
#[case::fzf("fzf")]
#[case::opentelemetry_go("opentelemetry-go")]
#[case::golang_benchmarks("golang-benchmarks")]
#[case::zerolog("zerolog")]
#[case::zap("zap")]
#[case::hugo("hugo")]
// Currently not producing results:
#[case::fuego("fuego")]
#[case::cli_runtime("cli-runtime")]
fn test_build_and_run(#[case] project_name: &str) {
    let temp_dir = setup_test_project(project_name).unwrap();

    // Mutex to prevent concurrent tests from interfering with CODSPEED_PROFILE_FOLDER env var
    static ENV_MUTEX: Mutex<()> = Mutex::new(());
    let _env_guard = ENV_MUTEX.lock().unwrap_or_else(|e| e.into_inner());

    let profile_dir = temp_dir.path().join("profile");
    unsafe { std::env::set_var("CODSPEED_PROFILE_FOLDER", &profile_dir) };
    if let Err(error) = crate::run_benchmarks(temp_dir.path(), ".") {
        panic!("Benchmarks couldn't run: {error}");
    }

    assert_results_snapshots(&profile_dir, project_name);
}
