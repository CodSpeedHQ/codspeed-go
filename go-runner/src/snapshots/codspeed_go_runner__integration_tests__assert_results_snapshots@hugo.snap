---
source: go-runner/src/integration_tests.rs
expression: results
---
[
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkAbsURL",
        "uri": "transform/urlreplacers/absurlreplacer_test.go::BenchmarkAbsURL",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAbsURLSrcset",
        "uri": "transform/urlreplacers/absurlreplacer_test.go::BenchmarkAbsURLSrcset",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkXMLAbsURL",
        "uri": "transform/urlreplacers/absurlreplacer_test.go::BenchmarkXMLAbsURL",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkXMLAbsURLSrcset",
        "uri": "transform/urlreplacers/absurlreplacer_test.go::BenchmarkXMLAbsURLSrcset",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkCSSEscaper",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkCSSEscaper",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkCSSEscaperNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkCSSEscaperNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkCSSValueFilter",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkCSSValueFilter",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkCSSValueFilterOk",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkCSSValueFilterOk",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkDecodeCSS",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkDecodeCSS",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkDecodeCSSNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/css_test.go::BenchmarkDecodeCSSNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkEscapedExecute",
        "uri": "tpl/internal/go_templates/htmltemplate/escape_test.go::BenchmarkEscapedExecute",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkHTMLNospaceEscaper",
        "uri": "tpl/internal/go_templates/htmltemplate/html_test.go::BenchmarkHTMLNospaceEscaper",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkHTMLNospaceEscaperNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/html_test.go::BenchmarkHTMLNospaceEscaperNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSRegexpEscaper",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSRegexpEscaper",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSRegexpEscaperNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSRegexpEscaperNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSStrEscaper",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSStrEscaper",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSStrEscaperNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSStrEscaperNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSValEscaperWithNum",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSValEscaperWithNum",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSValEscaperWithObj",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSValEscaperWithObj",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSValEscaperWithObjNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSValEscaperWithObjNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSValEscaperWithStr",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSValEscaperWithStr",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkJSValEscaperWithStrNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/js_test.go::BenchmarkJSValEscaperWithStrNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkSrcsetFilter",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkSrcsetFilter",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkSrcsetFilterNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkSrcsetFilterNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkStripTags",
        "uri": "tpl/internal/go_templates/htmltemplate/html_test.go::BenchmarkStripTags",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkStripTagsNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/html_test.go::BenchmarkStripTagsNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkTemplateSpecialTags",
        "uri": "tpl/internal/go_templates/htmltemplate/transition_test.go::BenchmarkTemplateSpecialTags",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkURLEscaper",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkURLEscaper",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkURLEscaperNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkURLEscaperNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkURLNormalizer",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkURLNormalizer",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkURLNormalizerNoSpecials",
        "uri": "tpl/internal/go_templates/htmltemplate/url_test.go::BenchmarkURLNormalizerNoSpecials",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkHash",
        "uri": "hugolib/pagesfromdata/pagesfromgotmpl_test.go::BenchmarkHash",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkListString",
        "uri": "tpl/internal/go_templates/texttemplate/parse/parse_test.go::BenchmarkListString",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkParseLarge",
        "uri": "tpl/internal/go_templates/texttemplate/parse/parse_test.go::BenchmarkParseLarge",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkVariableString",
        "uri": "tpl/internal/go_templates/texttemplate/parse/parse_test.go::BenchmarkVariableString",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkLoad",
        "uri": "config/allconfig/load_test.go::BenchmarkLoad",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkSortMap",
        "uri": "tpl/collections/sort_test.go::BenchmarkSortMap",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkWhereMap",
        "uri": "tpl/collections/where_test.go::BenchmarkWhereMap",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkWhereOps::eq",
        "uri": "tpl/collections/where_test.go::BenchmarkWhereOps::eq",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkWhereOps::like",
        "uri": "tpl/collections/where_test.go::BenchmarkWhereOps::like",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkWhereOps::ne",
        "uri": "tpl/collections/where_test.go::BenchmarkWhereOps::ne",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  },
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "BenchmarkTruncate::Plain_text",
        "uri": "tpl/strings/truncate_test.go::BenchmarkTruncate::Plain_text",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkTruncate::With_link",
        "uri": "tpl/strings/truncate_test.go::BenchmarkTruncate::With_link",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  }
]
